### DISCUSSION TITLE
How to manage several long running fibers with independent state?

### DISCUSSION THREAD

**Discussion by jontebol** [ioquatix/ioquatix#27] [Q&A] [answered]
Hi!

I'm developing a board game server / backend in Ruby/Rails. Basically, it receives events from clients via websocket and return events back. In between, each board needs to remember it's state and what it should do next. One way to do that is with fibers. Each board/session would have its own fiber which would yield as it encounters something that needs user input. Upon receiving the input, it would then resume were it was.

I had made a working implementation of this when I realized I had no good way to store all the fibers that would scale across multiple processes (because of no shared memory) or threads (because a fiber can't be resumed by a different thread). The only solution I've been able to figure out goes something like this:

event comes in with a board id
a "router" keeps track of in which process / thread the corresponding fiber lives and sends the event there
the router scales processes / threads as needed to handle the load (in other words, precisely what app servers do)

It seems this should be a rather common type of use case. And I'm relatively new to the whole concurrency thing. So I thought I would ask you:

1. Are my assumptions about fiber limitations in this use case correct?
2. Is there a different / better approach? *
3. If not, is there already something like this in the Async arsenal (or elsewhere)?

* I also have an implementation that is completely stateless, where the entire state is saved/restored (as well as a kind of "pointer" to where the execution should resume). This is actually the direction I'm going in since it seems much easier to scale / deploy. But I found the question of how to do this with fibers so interesting that I wanted to ask you if there is indeed a good way to do this.

**Comment by ioquatix** [comment, marked as answer]
The correct model for your logic is `update(current state, input) -> next state`.

Whether you do this per-request, use a database, etc, is totally up to you.

However, using fibers and WebSockets, you can avoid looking up the "current state" as each discrete connection will retain state over multiple updates. You need to be aware that users might make multiple connections and your "current state" cache might diverge. You'll want to use some kind of locking mechanism to ensure only one user per state cache.

Additionally, you should consider what happens if a WebSocket connection is lost. You should consider the WebSocket state as a cache, and make the board/session persist to a database. This ensures you have a robust system.

In summary:

1. Your assumptions about fibers are not wrong, but you can use the state handling to your advantage to avoid reloading the "current state" on every iteration of the loop.
2. Yes, you should have a stateless system + cached state per WebSocket connection + pessimistic locking.
3. You should check out live which provides the foundation for this: https://github.com/socketry/live

**Reply by jontebol** [reply]
Thanks for a great answer! Your description is basically what I'm doing, except for the cached state. I currently reload it on every message from the frontend, then run my loop (during which the state is "cached" I suppose you could say), return to the client and save the state again. Saving/restoring is cheap so it's not a big deal (currently at least), and I can then also use the db for locking.

If I understand the cache idea correctly, I should be fine doing that in a multiprocess / multithreaded environment as long as I don't cache paused Fibers like I was thinking of doing? And in general be careful that I don't have shared state that different threads may have access to? I may already have that... after the state loads, it's "cached" in class instance variables as the game runs. Not sure if that will be thread safe or if I will have to scope that state by a board/connection/session id or whatever. I guess I'll find out! 😅

Concurrency and parallelism is certainly not simple. Thanks for all the great work you're doing for the community btw! I've enjoyed listening to some talks of yours.

**Reply by ioquatix** [reply]
By far the simplest design is "share nothing". This means you do not share instances of a class between threads or fibers. This is also the most robust model. It's okay to have instance variables as long as the instance itself is not shared. Feel free to share more of your code (or a simple subset of the general code) and I'll provide more detailed feedback.
